### PR TITLE
Generator API + custom generators

### PR DESCRIPTION
**Pull request by zamotany**
See #419

### CODE REVIEW

File: src/commands/generate.js
Comment (toddw) [inline src/commands/generate.js:6]: Is this still a TODO?

File: src/generator/predefined/generator.js
Comment (toddw) [inline src/generator/predefined/generator.js]: There are a handful of places that use the word `modificator`. I think we really want `modifier`?

File: src/generator/predefined/reducer.js
Comment (toddw) [inline src/generator/predefined/reducer.js]: just mentioning one more time that I think all `modificator` should be `modifier`

File: test/generator/createTemplate.spec.js
Comment (toddw) [inline test/generator/createTemplate.spec.js]: The tests files are currently named `filename.spec.js`, the convention in the project is to use `filename.test.js`

**Comment by toddw** [comment]
It seems like all of the predefined generators are creating files with a blank line on top. I'm guessing this is because of how we are using
```
createTemplate`
/** … first line here… */
`;
```

It would be nice if we resolved that because I imagine a lot of people will do the same thing in their templates. Maybe we should check if the first line is blank then we remove it? What do you think?

File: src/generator/index.js
Comment (toddw) [inline src/generator/index.js]: It would be nice to all of the files that were generated along with the path to those files. Especially with custom generators which could potentially place files in un-obvious places.

**Comment by toddw** [comment]
Previously we were able to specify a folder when generating containers and components like:
```
gluestick generate component folderName/MyComponent
```

This would generate MyComponent in _src/components/folderName/MyComponent.js_ and _test/components/folderName/MyComponent.test.js_. This functionality is heavily used. It almost works in this PR but it throws an error and the output in the component needs to be updated slightly.

**Comment by toddw** [comment]
`gluestick generate reducer _reducer_name_` used to not only generate the reducers but it adds an export function on _src/reducers/index.js_ so that it is automatically picked up by the store in the GlueStick app. We definitely don't want to lose that functionality.

File: src/generator/predefined/component.js
Comment (toddw) [inline src/generator/predefined/component.js]: This should be the path `components/Todos` rather than `Todos`… I realize our previous template did not include `src/` in front but I would be okay if it does now since generators can generate files all over the place.

Comment (toddw) [reply src/generator/predefined/component.js]: but it shouldn't be a relative path like `../../src/components/Todos`

File: src/generator/predefined/component.js
Comment (toddw) [inline src/generator/predefined/component.js]: Would be nice to take advantage of our `components` alias. The import should be able to:
```
import ComponentName from "components/ComponentName";
```

We don't need the full relative path.

File: src/generator/predefined/container.js
Comment (toddw) [inline src/generator/predefined/container.js]: containers also has an alias so we should be able to do
`import ContainerName from "containers/ContainerName";`

**Review by toddw — CHANGES_REQUESTED** [review]
This is *seriously* great work! I'm very impressed and very happy with this PR. There are a few very minor changes that need to be make before we can merge but excellent work!

File: src/commands/generate.js
Comment (zamotany) [reply src/commands/generate.js:6]: Yeah, I'm fighting with commander to allow to pass options from cli to generator, but commander is   throwing exceptions all over the place.

**Comment by zamotany** [comment]
> It seems like all of the predefined generators are creating files with a blank line on top. 

Yeah it can be removed.

> Previously we were able to specify a folder when generating containers and components

Can do.

>  adds an export function on src/reducers/index.js

It's already implemented, see reducer generator

File: src/commands/generate.js
Comment (toddw) [reply src/commands/generate.js:6]: Should we create another ticket for tackling this or do you think it should block this PR?

**Comment by toddw** [comment]
> > adds an export function on src/reducers/index.js

> It's already implemented, see reducer generator

Can you verify that it works? I saw that but it didn't seem to work in a new project. I'll try again now

**Comment by zamotany** [comment]
Everything works like a charm.

File: src/commands/generate.js
Comment (zamotany) [reply src/commands/generate.js:6]: I will create new issue, since it more like a problem with commander. Besides it's not blocking, it's just an enhancement.

**Comment by toddw** [comment]
I've tested everything out and it works really well!!! So stoked about this!

**Pull request merged**
